README.md
# xmllint-wasm

libxml2's xmllint tool compiled to WebAssembly using Emscripten, to be used
in Node.js applications in environments where you can't or don't want to
depend on the native library.

This is a fork of Alan Zakai's amazing work at
[kripken/xml.js](https://github.com/kripken/xml.js)
(be sure to also check out his [blog post](http://mozakai.blogspot.com/2012/03/howto-port-cc-library-to-javascript.html) about
that exact project).
This fork continues the original build with some neat updates
as well as somewhat opinionated breaking changes.

Currently, the library only works in Node.js, but browser support could be
added if there's a demand for it. I'll probably stick to wasm builds only in
this repository, though, so if you need an asm.js build for browsers that
don't yet support wasm, you should probably go with the original project
instead.

## Overview of changes made to the original project

* libxml2 version is upgraded to v2.10.3
* The output is wasm instead of asm.js
* In addition to modern browsers with wasm support, works in Node.js 12 or later
* Library size is quite a bit smaller, the wasm file and wrapper js files
  weigh about 860K combined
* Allows normalization / formatting of the input XML in addition to validation
* There are some changes to the API, which is described in more detail
  below. Overall this project behaves more like a library that you'd call from
  a JS application, instead of like a command-line tool that xmllint normally is.

## Installation
```sh
npm i xmllint-wasm
```
The library uses Node.js [Worker threads](https://nodejs.org/api/worker_threads.html)
to isolate the Emscripten wrapper from your main process (so that
when it calls process.exit your whole server won't go down), which is
why Node >= 12 is required.

## API

See type definitions at [index.d.ts](./src/index.d.ts).

### Basic usage
```javascript
const {validateXML, memoryPages} = require('xmllint-wasm');

async function example() {
  const [myXMLFile, mySchemaFile, generalXmlXsdFile] = await Promise.all([
    fs.promises.readFile('./my-xml-file.xml', 'utf8'),
    fs.promises.readFile('./my-schema-file.xsd', 'utf8'),
    // Some other XSD file that our main schema my-schema-file.xsd happens to depend on
    fs.promises.readFile('./xml.xsd', 'utf8'),
  ])

  const validationResult = await validateXML({
    xml: [{
      fileName: 'my-xml-file.xml',
      contents: myXMLFile,
    }],
    // All the schema files that are required to validate the documents.
    // The main XSD should be first in the array, followed by its possible dependencies.
    schema: [mySchemaFile],
    // Optional: Files that the schema file depends on (xsd:import, xsd:include).
    // xmllint-wasm won't make any network requests or other IO to retrieve depdencicy files,
    // so these they must be loaded beforehand and passed in here.
    preload: [{
      fileName: 'xml.xsd',
      contents: generalXmlXsdFile,
    }],
    // Optional: Initial memory capacity in Web Assembly memory pages (1 = 6.4KiB) - 256
    // is minimum and default here (16MiB).
    initialMemoryPages: 256,
    // Optional: Maximum memory capacity, in Web Assembly memory pages. If not
    // set, this will also default to 256 pages. Max is 65536 (4GiB).
    // Use this to raise the memory limit if your XML to validate are large enough to
    // cause out of memory errors.
    // The following example would set the max memory to 2GiB.
    maxMemoryPages: 2 * memoryPages.GiB,
  });
  
  if (validationResult.valid) {
    console.log('There were no errors!')
  } else {
    console.warn(validationResult.errors);
  }
}

```
Giving explicit fileNames is optional (you can just pass the file contents
as string instead), but might help with mapping the correct error to a correct
file if you are validating multiple files at once. 

If you want the output to contain the (formatted) input XML, add
`normalization: 'format'` or `normalization: 'c14n'` to the options.

The return value is a Promise. Even though the xmllint command-line tool
returns with a non-zero exit code if the xml fails to validate, we
don't reject the Promise if there are validation errors as long as
the validation completes successfully. Unexpected errors, like
a syntax error in schema file, do reject the Promise.

The Promise resolved with a object like the following

```javascript

{
  valid: false,
  errors: [
    {
      // Raw output from running xmllint for
      // this particular error/line of the output
      rawMessage: "my-xml-file.xml:21: element quantity: Schemas validity error : Element 'quantity': [facet 'maxExclusive'] The value '1000' must be less than '100'.",
      // Error message without the location info
      message: "element quantity: Schemas validity error : Element 'quantity': [facet 'maxExclusive'] The value '1000' must be less than '100'.",
      // Error location info, if we managed to parse that from the output (null otherwise)
      loc: { fileName: 'my-xml-file.xml', lineNumber: 21 }
    }
  ],
  // Raw output string from running xmllint
  rawOutput: "my-xml-file.xml:21: element quantity: Schemas validity error ...",
  normalized: ""
}
```

### Usage in browsers

Check [./browser-demo/html-data](./browser-demo/html-data) index.html and
logic.js for a full example. Import the browser version, index-browser.js, as an
EcmaScript module and use like you would use the Node.js version.

Works out of the box in latest Chrome, Edge and Safari.  
Firefox, however, does not support Worker modules yet (tracked in issue
[1247687](https://bugzilla.mozilla.org/show_bug.cgi?id=1247687)), which would be
required by the implementation in this library.
To get Firefox to work, you'll need to use a bundler like [Parcel](https://parceljs.org/) or
Webpack to bundle the sources from modules to regular old scripts.

```javascript
import * as xmllint from './node_modules/xmllint/index-browser.mjs';
```

## Building xmllint from source

Clone the project (including the submodules with git `--recursive`) and build.  

### With Docker or Podman

```sh
  npm run build
```

### With locally installed Emscripten

[Install emscripten](https://emscripten.org/docs/getting_started/downloads.html#installation-instructions-using-the-emsdk-recommended)
and source their shell env.  
Finally, run the commands for Emscripten build

```sh
  ./script/clean
  ./script/libxml2
  ./script/compile
  ./script/test
```
There might also be some other system dependencies required for the build that are not listed here.


CHANGELOG.md
# Changelog

## 4.0.0

Add two new options, `initialMemoryPages` and `maxMemoryPages`, to control
how much memory the WebAssembly module allocates by default and is allowed
to allocate for the validation. 
[PR #16](https://github.com/noppa/xmllint-wasm/pull/13),
fixes [issue #8](https://github.com/noppa/xmllint-wasm/issues/8).  
Thanks [@MLSTRM](https://github.com/MLSTRM)!

The "**breaking change**" here is that we also raise the default maximum
memory cap from the old implicit 16MiB to double of that, 32MiB.  
With this new default, you may not even need to use the `maxMemoryPages` option
yourself, as 32MiB is probably already high enough for quite a lot of use cases.

In some _very_ memory constrained environments, this raised limit could impose
a problem if it causes the main process to run out of memory, so that's why
I'm calling it a breaking change. For most applications, though, this should
not be a problem. You can opt into the old limit by setting
```javascript
  maxMemoryPages: 16 * xmllint.memoryPages.MiB,
```

## 3.0.1

Fixes a broken build for Node in the previous version: Worker path
was defined relative to the index JS file, when Node actually resolves
relative Worker paths relative to the current working directory.
Fix by creating an absolute path with `path.resolve(__dirname, ..)`.

## 3.0.0

Should not contain breaking changes for most users, but there are
some internal file renames so I'm doing a major version bump just
in case someone was importing old files like index.js directly. 
`require('xmllint-wasm')` should still work as before.

* Upgrade libxml2 to 2.10.3
* Add browser support (fixes [#1](https://github.com/noppa/xmllint-wasm/issues/1)),
  thanks [@fhaftmann](https://github.com/fhaftmann) and [@th-we](https://github.com/th-we)!

## 2.2.0

No breaking changes to 2.0.0 - 2.1.0.

* Allow `schema` parameter to be omitted if `normalization` is set. i.e. allow
  formatting without doing validation to the XML ([PR #5](https://github.com/noppa/xmllint-wasm/pull/5)).

## 2.1.0

No known breaking changes to 2.0.0.

* Add support for xmllint option `--format` or `--c14n` via new config option
  `normalization` ([PR #3](https://github.com/noppa/xmllint-wasm/pull/3))
* Upgrade libxml2 version to 2.9.12 (fixes [#4](https://github.com/noppa/xmllint-wasm/issues/4) &
  [CVE-2021-3541](https://gitlab.gnome.org/GNOME/libxml2/-/commit/8598060bacada41a0eb09d95c97744ff4e428f8e))



src/index.js
'use strict';

/** @type {import("./index").MemoryPagesConstant} */
const memoryPages = {
	MiB: 16,
	GiB: 16384,
	defaultInitialMemoryPages: 256 /* 16MiB */,
	defaultMaxMemoryPages: 512 /* 32MiB */,
	max: 65536
};

function normalizeInput(fileInput, extension) {
	if (!Array.isArray(fileInput)) fileInput = [fileInput];
	return fileInput.map((xmlInfo, i) => {
		if (typeof xmlInfo === 'string') {
			return {
				fileName: `file_${i}.${extension}`,
				contents: xmlInfo,
			};
		} else {
			return xmlInfo;
		}
	});
}

function preprocessOptions(options) {
	const xmls = normalizeInput(options.xml, 'xml');
	const extension = options.extension || 'schema';

	validateOption(['schema', 'relaxng'], 'extension', extension);
	const schemas = normalizeInput(options.schema || [], 'xsd');
	const preloads = normalizeInput(options.preload || [], 'xml');
	const normalization = options.normalization || '';
	validateOption(['', 'format', 'c14n'], 'normalization', normalization);

	const inputFiles = xmls.concat(schemas, preloads);
	const args = [];
	schemas.forEach(function(schema) {
		args.push(`--${extension}`);
		args.push(schema['fileName']);
	});

	if (normalization) {
		args.push(`--${normalization}`);
	} else {
		// If no normalization is requested, we'll default to no output at all to "normalized" field.
		args.push('--noout');
	}

	xmls.forEach(function(xml) {
		args.push(xml['fileName']);
	});

	const opts = {
		inputFiles, args,
		initialMemory: options.initialMemoryPages || memoryPages.defaultInitialMemoryPages,
		maxMemory: options.maxMemoryPages || memoryPages.defaultMaxMemoryPages,
	};

	validateMemoryLimitOptions(opts);

	return opts;
}

function validationSucceeded(exitCode) {
	if (exitCode === 0) {
		return true;
	} else if (exitCode === 3 || exitCode === 4 /* validationError */) {
		return false;
	} else /* unknown situation */ {
		return null;
	}
}

function validateOption(allowedValues, optionName, actualValue) {
	if (!allowedValues.includes(actualValue)) {
		const actualValueStr = typeof actualValue === 'string' ? `"${actualValue}"` : actualValue;
		throw new Error(`Invalid value for option ${optionName}: ${actualValueStr}`);
	}
}

function validateMemoryLimitOptions({initialMemory, maxMemory}) {
	if (initialMemory < 0 || maxMemory < initialMemory || maxMemory > memoryPages.max) {
		throw new Error(
			'Invalid memory options.'
			+ ` Expected 0 < initialMemoryPages (${initialMemory}) <= maxMemoryPages (${maxMemory}) <= 4GiB (${memoryPages.max})`
		);
	}
}

function parseErrors(/** @type {string} */ output) {
	const errorLines = output
		.split('\n')
		.slice(0, -2);

	return errorLines.map(line => {
		const [fileName, lineNumber, ...rest] = line.split(':');
		if (fileName && lineNumber && rest.length) {
			return {
				rawMessage: line,
				message: rest.join(':').trim(),
				loc: {
					fileName,
					lineNumber: parseInt(lineNumber),
				}
			};
		} else {
			return {
				rawMessage: line,
				message: line,
				loc: null,
			};
		}
	}).filter(errorInfo => {
		// xmllint outputs "file.xml validates" for those files that are valid.
		const wasValid = !errorInfo.loc && errorInfo.rawMessage
			.trim()
			.endsWith(' validates');
		// don't list those files in errors list
		return !wasValid;
	});
}

/** @type {import("./index").validateXML} */
function validateXML(options) {
	const preprocessedOptions = preprocessOptions(options);
	var worker;

	return new Promise(function validateXMLPromiseCb(resolve, reject) {
		function onmessage(event) {
			// #ifdef browser
			var data = event.data;
			// #ifdef node
			var data = event;
			// #endif

			const valid = validationSucceeded(data.exitCode);
			if (valid === null) {
				const err = new Error(data.stderr);
				err.code = data.exitCode;
				reject(err);
			} else {
				resolve({
					valid: valid,
					normalized: data.stdout,
					errors: valid ? [] : parseErrors(data.stderr),
					rawOutput: data.stderr
					/* Traditionally, stdout has been suppressed both
					 * by libxml2 compile options as well as explict
					 * --noout in arguments; hence »rawOutput« refers
					 * only to stderr, which is a reasonable attribute value
					 * despite the slightly odd attribute name.
					 */
				});
			}
		}

		function onerror(err) {
			reject(err);
		}

		// #ifdef browser
		worker = new Worker(new URL('./xmllint-browser.mjs', import.meta.url), { type: 'module' });
		// #ifdef node
		const {Worker} = require('worker_threads');
		worker = new Worker(require('path').resolve(__dirname, './xmllint-node.js'));
		// #endif

		// #ifdef browser
		var addEventListener = worker.addEventListener.bind(worker);
		// #ifdef node
		var addEventListener = worker.on.bind(worker);
		// #endif

		addEventListener('message', onmessage);
		addEventListener('error', onerror);
		worker.postMessage(preprocessedOptions);
	}).finally(() => worker.terminate());
}

// #ifdef browser
export { validateXML, memoryPages };
// #ifdef node
module.exports.validateXML = validateXML;
module.exports.memoryPages = memoryPages;
// #endif
